python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1129,3 +1132,62 @@ def __new__(
     def value(self):
         """Alias of `entry_data`."""
         return self.entry_data
+
+
+T_NamespacedMetadataEntries = TypeVar(
+    "T_NamespacedMetadataEntries", bound="NamespacedMetadataEntries"
+)
+
+
+class NamespacedMetadataEntries(ABC, BaseModel, frozen=True):
+    """Extend this class to define a set of metadata fields in the same namespace.
+
+    Supports splatting to a dictionary that can be placed inside a metadata argument along with
+    other dictionary-structured metadata.
+
+    .. code-block:: python
+
+        my_metadata: NamespacedMetadataEntries = ...
+        return MaterializeResult(metadata={**my_metadata, ...})
+    """
+
+    @classmethod
+    @abstractmethod
+    def namespace(cls) -> str:
+        raise NotImplementedError()
+
+    def keys(self) -> AbstractSet[str]:
+        dictionary = self.dict()
+        return {
+            f"{self.namespace()}/{key}" for key in dictionary.keys() if dictionary[key] is not None
+        }
+
+    def __getitem__(self, key: str) -> Any:
+        return self.dict()[key.split("/", 1)[1]]
+
+    @classmethod
+    def from_dict(
+        cls: Type[T_NamespacedMetadataEntries], metadata_dict: Mapping[str, Any]
+    ) -> T_NamespacedMetadataEntries:
+        derived_dict = {
+            key.split("/", 1)[1]: value.value if isinstance(value, MetadataValue) else value
```

Comment:
seems like we should have one canonical spot for the `key.split("/", 1)[1]` logic

Reply:
👍